docs/Style.thy

```diff
@@ -219,6 +219,120 @@ lemma test_lemma3:
+  one style and you aren't doing a major rewrite, stick to the existing style.\<close>
+
+lemma
+  shows "\<lbrakk> A; B; C\<rbrakk> \<Longrightarrow>
+         D" \<comment> \<open>right-wrapping OK\<close>
+  and "\<lbrakk> A; B; C\<rbrakk>
+       \<Longrightarrow> D" \<comment> \<open>left-wrapping OK\<close>
+  oops \<comment> \<open>mixing styles: NOT OK\<close>
+
+text \<open>
+  Some operators and syntax only have ONE style. As seen in other sections:
+  * the `|` in `case`  never appears on the right
+  * `;` is always on the right when wrapping lists of assumptions
+  * `shows .. and ... and` wraps with `and` on the left
+  * commas and colons, owing to our natural language origins, always end up on the right\<close>
+
+lemma
+  shows
+    "\<lbrakk> A
+     ; B \<rbrakk> \<comment> \<open>wrong: always on right\<close>
+     \<comment> \<open>ok: can be either\<close>
+     \<Longrightarrow> C" and \<comment> \<open>wrong: `shows/and` only on left!\<close>
+    "D"
+  and "E" \<comment> \<open>ok: on left\<close>
+proof -
+  have "True \<and> True"
+    by (rule conjI,
+        blast,
+        blast) \<comment> \<open>ok\<close>
+  have "True \<and> True"
+    by (rule conjI
+        , blast
+        , blast) \<comment> \<open>NOT OK: commas go on right\<close>
+  have "True \<and> True"
+    by (rule conjI;
+        blast) \<comment> \<open>ok\<close>
+  have "True \<and> True"
+    by (rule conjI
+        ; blast) \<comment> \<open>also ok, can be either, but prefer ; on right if it's a simple method
+                    application\<close>
+  oops
+
+text \<open>
+  The general principle of "nothing indented less than what it belongs to" is in effect for both
+  wrapping styles. Remember, the goal of the exercise is to make it as clear to read for others as
+  you can. Sometimes, scanning the left side of the screen to see the overall term can help,
+  while other times putting @{text \<Longrightarrow>} on the right will save space and prevent subsequent lines
+  from wrapping.\<close>
+
+text \<open>
+  Inner-syntax indentation is not automatable in the general case, so our goal is to help
+  ease of comprehension as much as possible, i.e.
+  @{term "A \<and> B \<or> C \<longrightarrow> D \<or> E \<and> F"} is bearable if A..F are short, but if they are large terms,
+  please avoid doing either of these:\<close>
+
+term "
+  A \<and>
+  B \<or>
+  C \<longrightarrow>
+  D \<or>
+  E \<and>
+  F" \<comment> \<open>avoid: requires building a parse tree in one's head\<close>
+
+term "
+  A
+  \<and> B
+  \<or> C
+  \<longrightarrow> D
+  \<or> E
+  \<and> F" \<comment> \<open>can be marginally easier to scan, but still avoid due to mental parsing difficulty\<close>
+
+text \<open>Instead, help out the reader like this:\<close>
+
+term "
+  A \<and>
+  B \<or>
+  C \<longrightarrow>
+    D \<or>
+    E \<and>
+    F"
+
+term "
+  A
+  \<and> B
+    \<or> C
+  \<longrightarrow> D
+     \<or> E
+       \<and> F"
```

Comment:
I'm struggling to give an alternative suggestion, but these two examples don't quite seem right to me. At the very least, to be consistent should the right-wrap style have more indentation around the nested conj's and disj's?

Reply:
I think both could be permitted. Depends on how used you are to reading `/\` and `\/` as a list and how clear the operator binding is. If you want to emphasise you'd probably write

```
term "
   A /\
   B \/
   C -->
     D \/
       E /\
       F"
```

I don't think I'd write

```
term "
   A /\
   B \/
     C -->
       D \/
         E /\
         F"
```

because that would indicate `A /\ B \/ (C --> D \/ E /\ F)` instead of `(A /\ B \/ C) --> (D \/ E /\ F)`

I guess you could go with:

```
term "
   A /\
   B \/
     C
   -->
   D \/
     E /\
     F"
```
 if you really wanted to emphasise the precedence. But my normal preference would be to keep it flat unless it becomes unclear.

Reply:
So, uh, what do I put down for the "do" example for right-wrap?
I think the last one least (where `-->` isn't indented, and neither is `D` after the implication). The difference is, I think, that if you're dumping the operators on the far right with a ragged edge, the extra indentation isn't really going to help for the leading `\/`.
I like Gerwin's first example the most here, but I didn't write it myself because it's hard to quantify why I would have written it that way.

For the left-wrap, it makes more sense to show the spine. I guess you could flatten it as well, but with long terms you'd probably indent it, but flatten it if it's on one line.

Reply:
The second example (operators on left) is also a bit troubling, as @corlewis notes:
```
A
/\ B
   \/ C
--> D
    \/ E
       /\ F
```
...because it suggests `(A /\ (B \/ C)) --> (D \/ (E /\ F))`, which gives the wrong precedence between `/\` and `\/` in the antecedent. The issue seems to occur when you have a higher precedence operator (`/\`) before a lower precedence operator (`\/`), regardless of whether you're putting operators on the left or right. We naturally want to indent *below* something, not above it.

So unless we want to indent the `A /\ B` above the `\/ C`, I suggest we make it acceptable to flatten the LHS as long as it's safe for the reader to associate to the left without paying much attention to the operators.

For operators-on-the-right, I think that means you get @lsf37's first example:
```
A /\
B \/
C -->
  D \/
    E /\
    F
```

For operators-on-the-left, I think it would be: 
```
A
/\ B
\/ C
--> D
    \/ E
       /\ F
```

I tried to make more precise rules for which lines should be indented, but it got tricky, so I think it's probably best to keep it imprecise, i.e. indent as a hint to the reader that they need to break left associativity.

Does that make any sense?

Reply:
Thanks Matt!
Yes, it makes sense, and as you both say I've made a booboo (in fact, did the thing I said not to do, which is mislead the reader). I thought about it for a bit, and how to phrase a rule that we'd want in this case... and I'm also not clear on what that rule should be. As Matt says, this might be a place to not make a rule.
I'm also happy with both the right and left examples that Matt gave to go in as examples, @corlewis @lsf37 are you two in agreement?

Reply:
Yes, agreement. A high-level guideline we could put into text form is to say that the layout should not mislead about operator precedence.